### PR TITLE
Update purefa_host module

### PR DESCRIPTION
**Pull request by sdodsley**
##### SUMMARY
Update purefa_host module.
Add support for specifying the LUN ID of a volume when connecting to a host.
Add support for updating FC WWNs and iSCSI IQNs on a host
Add support for changing host personality
Add support for NVMeF hosts
Fix support for adding volumes to existing hosts - broke somewhere in an update...

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
storage/purestorage/purefa_host.py

### CODE REVIEW

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 1]
```diff
@@ -35,21 +35,32 @@
...
+    - List of NQNs of the host if protocol is nvmef or mixed.
+    version_added: '2.8'
   volume:
     description:
     - Volume name to map to the host.
+  lun:
+    description:
+    - LUN ID to assign to volume for host. Must be unique.
+    - If not provided the ID will be automatically assigned.
+    - Range for LUN ID is 1 to 256.
```
Comment (sile16) [inline lib/ansible/modules/storage/purestorage/purefa_host.py]: LUNs 1-4095 for private or shared connections, is what the array supports.

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 2]
```diff
@@ -196,25 +311,31 @@ def main():
...
 
+    api_version = array._list_available_rest_versions()
+    if module.params['nqn'] is not None and NVMEF_API_VERSION not in api_version:
+        module.fail_json(msg='NVMeF protocol not supported. Please upgrade your array.')
     state = module.params['state']
     protocol = module.params['protocol']
     array = get_system(module)
     host = get_host(module, array)
-
+    if module.params['lun'] and not 1 <= module.params['lun'] <= 256:
```
Comment (sile16) [inline lib/ansible/modules/storage/purestorage/purefa_host.py]: LUNs 1-4095 for private or shared connections

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 3]
```diff
@@ -127,67 +151,158 @@
...
+                array.set_host(module.params['host'],
+                               addwwnlist=module.params['wwns'])
+            except:
+                module.fail_json(msg='FC WWN additiona failed.')
+
+
+def _connect_new_volume(module, array, answer=False):
+    """Connect volume to host"""
+    api_version = array._list_available_rest_versions()
+    if AC_REQUIRED_API_VERSION in api_version and module.params['lun']:
```
Comment (sile16) [inline lib/ansible/modules/storage/purestorage/purefa_host.py:208]: The lun param has always been there.  Shouldn't need to check for ActiveCluster support just to send the vol LUN ID.

Comment (sdodsley) [reply lib/ansible/modules/storage/purestorage/purefa_host.py:208]: So it has. My mistake. Removing it.

File: lib/ansible/modules/storage/purestorage/purefa_host.py (hunk 1 above)
Comment (sdodsley) [reply lib/ansible/modules/storage/purestorage/purefa_host.py]: Was sticking to 256 as a limit as not all OSes can support values as high as 4095. I can change that and add a warning.

File: lib/ansible/modules/storage/purestorage/purefa_host.py (hunk 2 above)
Comment (sdodsley) [reply lib/ansible/modules/storage/purestorage/purefa_host.py]: As above

**Comment by sile16** [comment]
shipit

**Review by sile16 — APPROVED** [review]

**Pull request merged**